### PR TITLE
Add PPO with clipping objective

### PR DESCRIPTION
**Pull request by magnusja**

### CODE REVIEW

**Comment by pat-coady** [comment]
Hey Magnus,

Thanks a lot. I think I read in your comments that training performance wasn't very good with clipping. Do you still think it is worthwhile to add to the code? Let me know your thinking. Is it possible that the paper use another method in combination with clipping that made it perform well?

Pat

**Comment by magnusja** [comment]
Hey Pat,

so I first thought it converges slower, but I am able to get >3500 reward for hopper environment after ~10k episodes. So that should be fine I think.
I would include it because that is actually the method they consider as best in the PPO paper.

There are some tweaks which seem to be better for more complex environments.
1. Lowering learning rate after sampling (very first step of optimization) because PPO clipping objective has no effect there. the PG ratio is always 1 at the first step of optimization. After that the clipping objective is "active" and learning rate can be increased.
2. Using an adaptive learning rate similar to [1]. They mention in the paper that they use that for RoboSchool I think.

[1] https://github.com/magnusja/homework/blob/master/hw4/main.py#L353

**Comment by pat-coady** [comment]
Is this the paper you're referring to:

https://arxiv.org/pdf/1707.06347.pdf

**Comment by magnusja** [comment]
Exactly!

**Comment by pat-coady** [comment]
Ok, will review tomorrow. Thanks!

**Comment by pat-coady** [comment]
Very sorry for the delay. Thanks a lot for the improvement. Merged with master.

**Comment by pat-coady** [comment]
If you are still interested, can you resubmit PR? I merged with another change that caused some problems. So I had to revert both. I can't seem to find just your change to merge.

**Comment by magnusja** [comment]
What happened? I am utterly confused. Github says that you pushed into MY repo: https://github.com/magnusja/ppo

How is that even possible?

**Comment by pat-coady** [comment]
I'm far from GitHub expert, but I wouldn't think that's possible. Maybe you pulled my changes before I reverted?

**Comment by magnusja** [comment]
I just checked that and apparently one has write access to every forked repo, which is a little weird.
So for some reason, you pushed into my repo. It was certainly not me :D

I will try to revert and open a new PR.

**Comment by pat-coady** [comment]
Sorry about that! I have no idea how I might have pushed a change while
merging something into my repo. Odd.

**Comment by magnusja** [comment]
Hmm ok unfortunately my commit is already in your history, hence I cannot open a new PR: https://github.com/pat-coady/trpo/compare/master...magnusja:master

**Comment by pat-coady** [comment]
Ok, I'll do a little stack overflow search and see how I can merge that
commit.

**Comment by pat-coady** [comment]
Ok, I was able to grab that commit. Everything is all set. Thanks for your
contribution. And sorry for the trouble.
